implementations/rust/ockam/ockam_transport_tcp/src/portal/inlet_shared_state.rs

```diff
@@ -1,25 +1,131 @@
-    is_paused: bool,
-    // Starts with 0 and increments each time when inlet updates the route to the outlet
-    // (e.g. when reconnecting), this will allow outlet to figure out what is the most recent
-    // return_route even if messages arrive out-of-order
-    route_index: u32,
+    routes: Arc<SyncRwLock<Vec<(String, InletRouteState)>>>,
+    // To notify the listener that a new route is available
+    notify: Arc<Notify>,
 }
 
 impl InletSharedState {
-    pub fn create(ctx: &Context, route: Route, is_paused: bool) -> Result<Self> {
+    /// Add a new route to the shared state and return true if the route was added successfully.
+    /// If the route already exists, the function will return false.
+    pub fn try_add_route(&self, key: String, route_state: InletRouteState) -> bool {
+        let mut guard = self.routes.write().unwrap();
+        if guard.iter().any(|(k, _)| k == &key) {
+            false
+        } else {
+            guard.push((key, route_state));
+            self.notify.notify_waiters();
+            true
+        }
+    }
+
+    /// Adds a new route to the shared state and return an error if the route already exists.
+    pub fn add_route(&self, key: String, route_state: InletRouteState) -> Result<()> {
+        if !self.try_add_route(key, route_state) {
+            Err(ockam_core::Error::new(
+                Origin::Channel,
+                Kind::AlreadyExists,
+                "Route already exists",
+            ))
+        } else {
+            Ok(())
+        }
+    }
+
+    /// List all route keys
+    pub fn list_all_route_keys(&self) -> Vec<String> {
+        let guard = self.routes.read().unwrap();
+        guard.iter().map(|(k, _)| k.clone()).collect()
+    }
+
+    /// Removes a route given its key
+    pub fn remove_route(&self, route_key: &str) {
+        let mut guard = self.routes.write().unwrap();
+        guard.retain(|(k, _)| k != route_key);
+    }
+
+    /// Choose one random active route from the shared state, if no active routes are found, wait
+    /// until a new route is added or an existing route is unpaused.
+    pub async fn choose_active_route(&self) -> InletRouteState {
+        loop {
+            {
+                let guard = self.routes.read().unwrap();
+                let active_routes: Vec<&InletRouteState> = guard
+                    .iter()
+                    .filter(|(_, route)| !route.is_paused())
+                    .map(|(_, route)| route)
+                    .collect();
+
+                if !active_routes.is_empty() {
+                    break active_routes
+                        .choose(&mut rand::thread_rng())
+                        .copied()
+                        .cloned()
+                        .unwrap();
+                }
+            }
+            // let's wait until something changes and try again
+            debug!("No active route found for the inlet, waiting for a connection");
+            self.notify.notified().await;
```

Comment:
I think for both regular and privileged portals `choose_active_route` would better return an error instead of waiting for a connection. WDYT?

Reply:
I think both are acceptable, and it could be an option:
By waiting for it, the connection connects to the TCP inlet, starts sending data, and then it waits until a route is established.
The alternative being a direct rejection of the connection.

The former has the advantage of avoiding the client polling time, and manage to resume the connection as soon as it's available.
The latter has the advantage of better mimicking the availability of the service.